### DISCUSSION TITLE
acces ROM waveforms when installed on E2 synth ?

### DISCUSSION THREAD

**Discussion by chaocrator** [bangcorrupt/hacktribe#77] [Ideas]
would it be possible at some point to acces ROM waveforms on E2 synth ?

at least drums/percussions in E2 synth are perfectly usable.

also, E2 synth owners have no option to export factory samples, so after Hacktribe installation we get only VA/FM oscillators, and that's all. we have to ask for factory waveforms (from E2S) elsewhere if we need them.

**Comment by bangcorrupt** [comment]
Yes, this will be possible in the future.  When the PCM data is loaded from the flash, it's decompressed and sent to DSP RAM one word at a time. I can hijack this to copy PCM to CPU RAM, then dump it over sysex as a single blob. If I can get a debugger attached to the DSP then I may be able to fix the issue with sample saving and have the factory samples loading like normal.

If you're short of samples until then, check out the [carved beats set](https://github.com/bangcorrupt/hacktribe-content/tree/master/carved_beats_set) by @BKLronin.

**Reply by maks** [reply]
What a great collection of samples! Thanks for making them available @BKLronin !

**Reply by BKLronin** [reply]
Thank you very much. They were exclusively recorded and synthesised for hacktribe.

**Comment by maks** [comment]
@bangcorrupt I've just reread this and thought about the implications: would this mean that its in theory possible that you would be able to in the future modify the firmware to send/receive sample data via Midi sysex?

**Reply by bangcorrupt** [reply]
That's the plan #65.  I'll probably work on this when I next look at multi-samples.  At some point I want to remove the PCM data from the flash entirely and use the space for larger pattern files.

**Reply by j-grosse** [reply]
Hi @bangcorrupt , I have the E2 - Synth version (not the E2S - Sampler Version) with the normal, non-hacked Synth Firmware version 2.02. and I want to get the E2 ROM PCM WAV Factory Sounds. Then I want to make my E2 a Hacktribe (= enhanced E2S) but with all the E2 ROM PCM WAV Samples. And then in addition I want to add my own Samples and the E2S Factory Samples. Is this possible? 
I tried the dump_extracted_pcm.py python script as described here: https://github.com/bangcorrupt/hacktribe/wiki/Dump-Synth-PCM-Samples
But I get the error that the OS version is wrong and that no ports are found. I used a MacBook with MacOS Catalina and connected my Electribe Synth version via USB cable. I tried with python 2.7 and python 3.x

Or do I have to first install the synth-hacked-SYSTEM.VSB and before restarting the E2 (so before It becomes a hacktribe) use the dump_extracted_pcm.py python script to extract the E2 ROM PCM WAV Sounds from RAM?

Thanks in advance!
Jay

**Reply by bangcorrupt** [reply]
You need to patch the factory firmware using the `synth-pcm-dump` branch then install it, reboot and run the script.  It should give you the synth PCM as one file.  Then you can patch the factory firmware using the `main` or `experimental` branch to install Hacktribe.

> error that the OS version is wrong and that no ports are found

Can you paste the output?

All the scripts are Python 3.

**Reply by j-grosse** [reply]
@bangcorrupt @AndrewSosna @maks  I fixed a problem with the midi driver on my MacBook. But I still cannot get the PCM Sounds. The port name is unknown.

What I did:
- ran the tutorial script to create the file “synth-hacked-SYSTEM.VSB” from the hack tribe synth-pcm-dump branch via:
https://github.com/bangcorrupt/hacktribe-tutorial/tree/synth-pcm-dump
- renamed it to SYSTEM.VSB and put it onto the SD card into the folder “/KORG/electribe/System“
- did the firmware update with it on my grey E2 synth
- before restarting the E2 I removed the SD card
- restarted the E2 and connected it to the my Macbook via USB cable
- installed KORG USB driver 
- uninstalled it afterwards because I have MacOS Catalina installed on my MacBook, which has its own usb driver to connect to the E2. See https://www.korg.com/us/news/2019/112902/

- cloned the hack tribe repo from Github, went into the “hack tribe” folder
- installed python 3 globally
- installed dependencies via “pip3 install mido python-rtmidi construct” in the cloned folder “hacktribe”
- ran “python3 scripts/dump_extracted_pcm.py”

- received the following terminal output:

~/Projects/hacktribe on branch synth-pcm-dump! ⌚ 8:41:46
$ python3 scripts/dump_extracted_pcm.py 

DEBUG:root:Initialise SysEx
Traceback (most recent call last):
  File "/Users/admin/IdeaProjects/hacktribe/scripts/dump_extracted_pcm.py", line 8, in <module>
    e = E2Sysex()
  File "/Users/admin/IdeaProjects/hacktribe/scripts/e2sysex.py", line 31, in __init__
    self.inport = mido.open_input(port)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/mido/backends/backend.py", line 95, in open_input
    return self.module.Input(name, **self._add_api(kwargs))
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/mido/ports.py", line 165, in __init__
    BasePort.__init__(self, name, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/mido/ports.py", line 90, in __init__
    self._open(**kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/mido/backends/rtmidi.py", line 133, in _open
    self.name = _open_port(self._rt, self.name, client_name=client_name,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/mido/backends/rtmidi.py", line 101, in _open_port
    raise OSError(f'unknown port {name!r}')
OSError: unknown port 'electribe2 sampler electribe2 s'

**Reply by bangcorrupt** [reply]
The ports may get different names on Mac.  Try running this in a python console:

    import mido
    mido.get_input_names()

It should print a list of MIDI port names as `mido` sees them.

Edit line 21 in `scripts/e2sysex.py` to include the correct port name:

    def __init__(self, port='your electribe port name', midi=None):

**Reply by F-Thom** [reply]
Yesterday I cam to the same issue and I found out that it may be another reason that it doesn't work.
Disclaimer: I'm quite a beginner and use codespace for the first time.
I wanted to patch the E2 stock firmware and noticed that patching the firmware didn't work, it patched with the Hacktribe firmware systematically. The reason: the line 47 of hacktribe-tutorial.sh is cloning the main branch and not synth-pcm-dump, even if i choose h right branch before opening the codespace template. Is there here an issue?
Thanks in advance and thanks for everything you've already done on the firmware !

**Reply by bangcorrupt** [reply]
Good catch.  It looks like template repos always create a codespace from the main branch.  As a workaround for now, either click `Code > Codespaces > Create a codespace on synth-pcm-dump` (after switching branches), or edit the clone command in the script to add `-b synth-pcm-dump` and run it again from the terminal using `./hacktribe-tutorial.sh`.

https://github.com/bangcorrupt/hacktribe-tutorial/issues/2

**Reply by F-Thom** [reply]
Thank you very much for your quick reaction, I could achieve the first step: E2S with E2 sounds.
As next I updated the e2sysex.py script with a very straight implementation (to be improved later possibly):
class E2Sysex:

    def __init__(self, port='electribe2 sampler 1 SOUND', midi=None):
        logging.debug('Initialise SysEx')
       
        mido.get_input_names()
        if midi is not None:
            self.midi = midi
            self.inport = self.midi.inport
        else:
            # ADD - try, catch
            logging.debug('Port Electribe not found')
            port = 'electribe2 sampler 1 PAD/KNOB 0'
            self.inport = mido.open_input(port)

        mido.get_output_names()
        if midi is not None:
            self.midi = midi
            self.outport = self.midi.outport
        else:
            # ADD - try, catch
            logging.debug('Port Electribe not found')
            port = 'electribe2 sampler 1 SOUND 1' 
            self.outport = mido.open_output(port)

        self.global_channel, self.id, self.version = self.search_device()
   
        logging.info('Found electribe')
        logging.info('Global channel ' + str(self.global_channel))
        logging.info('Firmware version ' + self.version)
    
        self.sysex_head = [0x42, 0x30 + self.global_channel, 0x00, 0x01, self.id]
        
        Where:
        'E2 Name for your input' is the answer of mido.get_input_names()
        'E2 Name for your output' is the answer of mido.get_output_names()
So yes, the names on midi are not the same for input and output, at least for me but I think this may be more generally the case.

Now I'm blocked at the RAM read (read_cpu_ram) where the SysEx is supposed to be send but no answer comes...
SysEx works generally bidirectionally because the Electribe firmware name can be read but the read ram command doesn't work further...

**Reply by bangcorrupt** [reply]
It will take a long time to dump the PCM data over sysex, it may look like it is doing nothing.   I will try to test this later.

Can you post the log output from running `dump_extracted_pcm.py` ?

Do other sysex functions work?  You could test this by editing `dump_extracted_pcm.py` to get patterns and global settings before reading the PCM data:

    ...
    pattern = e.get_current_pattern()
    global = e.get_global()
    pcm_decomp = e.read_cpu_ram(0xc3000000, 0xf9b2ea)
    ...

**Reply by F-Thom** [reply]
The log is following:
extracted_pcm.py
DEBUG:root:Initialise SysEx
DEBUG:root:Port Electribe not found => probably my fault
DEBUG:root:Port Electribe not found => probably my fault too
INFO:root:Found electribe => search_device is successful so SysEx generally work
INFO:root:Global channel 0
INFO:root:Firmware version 2.2
DEBUG:root:Getting decompressed PCM from CPU RAM

The function  e.get_current_pattern() and e.get_global() don't work either I think the RAM read function is not properly answering.
This is no urgent topic at all, I took the time to check this today but it's more than a year that I didn't try to uprade my Elecribe

**Reply by alexreisinger** [reply]
Hey I would like to bump this as I am stuck on this exact step.
I patched the firmware and got everything running on my E2 synth.
Connected the usb cable and checked mido.get_input_names() and mido.get_output_names() 
and they return 'electribe2 sampler 0' for the input and 'electribe2 sampler 1' for the output so I adapted the script to use those resepctively:

e2sysex.py:
```python
def __init__(self, input_port='electribe2 sampler 0', output_port='electribe2 sampler 1', midi=None):
        logging.debug('Initialise SysEx')
        
        if midi is not None:
            self.midi = midi
            self.inport = self.midi.inport
            self.outport = self.midi.outport
        else:
            self.inport = mido.open_input(input_port)
            self.outport = mido.open_output(output_port)
``` 

when running the script I am stuck at the same spot as @F-Thom (maybe you got more progress by now?).
If I leave the ports standard I also get the error of the Ports not being found.

output of the script: 
PS C:\Users\alexr\hacktribe> python .\scripts\dump_extracted_pcm.py
DEBUG:root:Initialise SysEx
INFO:root:Found electribe
INFO:root:Global channel 0
INFO:root:Firmware version 2.2
DEBUG:root:Getting decompressed PCM from CPU RAM
DEBUG:root:Called read_cpu_ram
DEBUG:root:Sending message

That is where everything blocks. I cant seem to get a response from the other SysEx functions you mentioned, the e.get_current_pattern() and so on - it just runs indefinitley.

Using Windows 10 and VSCode for that matter. Hopefully you can help and I am happy to provide any further information if necessary! Thanks for the effort in the whole project, looking forward to testing it with the PCM Samples.

**Reply by bangcorrupt** [reply]
I don't know why this isn't working for people.  I have just tested and it worked for me, taking a little over a minute and giving me a 16MB wav file of the PCM data.  I did get a 'Resource temporarily unavailable' message, possibly due to flooding the input with a large sysex message.  Maybe this is handled differently on Windows and causes an error.  I will try to test with Windows, but it might not be soon.  

```
[user@host hacktribe] $ time python scripts/dump_extracted_pcm.py
DEBUG:root:Initialise SysEx
INFO:root:Found electribe
INFO:root:Global channel 0
INFO:root:Firmware version 2.2
DEBUG:root:Getting decompressed PCM from CPU RAM
DEBUG:root:Called read_cpu_ram
DEBUG:root:Sending message

MidiInAlsa::alsaMidiHandler: unknown MIDI input error!
System reports: Resource temporarily unavailable
DEBUG:root:Received response
DEBUG:root:Decoding binary
DEBUG:root:Writing pcm-dump.wav

real    1m9.706s
user    0m10.973s
sys     0m0.904s
```

**Reply by bangcorrupt** [reply]
Can you check you have 'python-rtmidi' installed, not 'rtmidi-python' (these are different packages)?

Could you also check which backend mido defaults to on your system?

In a python console:
```
>>> import mido
>>> mido.backend
<backend mido.backends.rtmidi (not loaded)>
```

**Reply by F-Thom** [reply]
> I don't know why this isn't working for people. I have just tested and it worked for me, taking a little over a minute and giving me a 16MB wav file of the PCM data. I did get a 'Resource temporarily unavailable' message, possibly due to flooding the input with a large sysex message. Maybe this is handled differently on Windows and causes an error. I will try to test with Windows, but it might not be soon.
> 
> ```
> [user@host hacktribe] $ time python scripts/dump_extracted_pcm.py
> DEBUG:root:Initialise SysEx
> INFO:root:Found electribe
> INFO:root:Global channel 0
> INFO:root:Firmware version 2.2
> DEBUG:root:Getting decompressed PCM from CPU RAM
> DEBUG:root:Called read_cpu_ram
> DEBUG:root:Sending message
> 
> MidiInAlsa::alsaMidiHandler: unknown MIDI input error!
> System reports: Resource temporarily unavailable
> DEBUG:root:Received response
> DEBUG:root:Decoding binary
> DEBUG:root:Writing pcm-dump.wav
> 
> real    1m9.706s
> user    0m10.973s
> sys     0m0.904s
> ```

First of all, thank you very much for coming back to us and taking time to check this topic again. 
I didn't check further after my last message and I can try when I'll have time again. I'm not in a hurry. I can report whenever I've some news.

**Reply by alexreisinger** [reply]
Ok I finally got time to check on that again.
I got it to work, I now have a 16MB wav file.
On Windows and on my Linux VM inside Windows I always got stuck at: DEBUG:root:Sending message, but now on an M3 MacBook Air I got it to work flawlessly.
I just had to adapt the Input and Output ports in the e2sysex.py to 
input_port='electribe2 sampler PAD/KNOB'
output_port='electribe2 sampler SOUND'

The script however does not display the time stats that your log stated, it just stops at Writing pcm-dump.wav.

The result of the mido.backend is:
>>> import mido
>>> mido.backend
<backend mido.backends.rtmidi (loaded)>


Thanks again for your work, and I will now continue the workflow. If anyone has any questions maybe I can help a bit on that part!

**Comment by drmindflip** [comment]
I joined up here to ask this! I'm an E2 Synth user and I mostly use it for making ambient sequences and textures - I literally never use the 90% of the stock sounds (OSC types 001- 325 never get a look in) but I do get a lot of use out of 326-428 (the Synth and Instrument categories)

My ideal would be to keep those, trash the rest, get those lovely new synth oscs that you've added, and maybe eventually add a few samples. Is this ever likely to be possible?

Thanks for all your inspiring work

**Reply by bangcorrupt** [reply]
Dumping the raw PCM data from flash will definitely happen.  The pianos and some other instruments are multi-sampled, I can probably get this working at some point.  The Osc Edit knob applies a drive effect, I hope I can get this working one day.

**Reply by drmindflip** [reply]
Thanks for your reply and awesome work. My dream would be to keep the E2 Synth basically as-is, but to replace most of OSCs 001-325 with some more versatile one-shots and sound design bits. I actually love the limitations in using 326-428 (I always seem to go blank when trying to prepare material for a sample-based workflow) but a few more instrument samples wouldn't go astray.

**Reply by AndrewSosna** [reply]
Hi @bangcorrupt . Just wanted to ask if you had a chance to check the possibility of getting those E2Synth samples? For some weird reason I like them and would like to keep them or at least some.

**Reply by bangcorrupt** [reply]
Hi, just had a look at this.  I can make a separate patch to decompress the PCM and a script to dump it over SysEx.  Should be ready sometime tomorrow, if it works first time...

**Reply by AndrewSosna** [reply]
Awesome, that would be great.

**Reply by bangcorrupt** [reply]
I pushed a new branch with a patch and script for dumping synth PCM.  Slicing it up and naming the samples correctly is more work than I thought, so for now it's just one long block with all the samples together.  There are some notes in the wiki [here](https://github.com/bangcorrupt/hacktribe/wiki/Dump-Synth-PCM-Samples).

**Reply by AndrewSosna** [reply]
Great, I'll check tomorrow, thank you.

**Reply by AndrewSosna** [reply]
@bangcorrupt , I did some testing and either I did something wrong and resulting WAV file is incorrect, or I incorrectly assumed that it should be playable, just have all the sounds playing in queue. I'll post my process just in case

Environment: macOS Ventura 13.1, Python 3.7.9, MIDI-USB, E2 Synth (blue), branch synth-pcm-dump
I've patched the original sampler firmware 2.2, modifying header in the process, updated the firmware, Electribe starts as a Sampler.
Running the patch failed at first:

- absent e2_syx_codec.py, I copied it from your hacktribe-editor repo
- missing mido and python-rtmidi, install them
- the script could not find the port named `electribe2 sampler electribe2 s`, so I've checked what are available on my device and had to set
```
            self.inport = mido.open_input('electribe2 sampler PAD/KNOB')
            self.outport = mido.open_output('electribe2 sampler SOUND')
```
There was no in/out ports.

After that the dump script worked, produced e2-synth-pcm.wav but the file is just hissing, consists mostly of "FFFFFFFF" and a WAV header and occasional data that seems to be random. I've repeated the dump a couple of times expecting to get binary equal WAV files every time but sometimes the files were different. Just a note, I changed volume of my Electribe between dump sessions.

I know that this WAV should not be used as is and must be split first, but still, should it be playable or maybe I did something wrong?

**Reply by bangcorrupt** [reply]
Assuming the samples play correctly when the synth firmware is installed, the dumped file should play the same sounds.  If it is mostly set bits then it has probably dumped an empty section of RAM.  Changing the volume won't change the data in RAM, but each reboot could if some areas are never initialised.  There is probably a bug with the patch preventing the PCM data writing to the correct address.  Can you email me the patched firmware you have installed and the wav file you dumped?

**Reply by AndrewSosna** [reply]
Thanks, sent you an email

**Reply by bangcorrupt** [reply]
The firmware you sent me matches the one I tested and when I run it I get the samples dumped correctly.  The audio files you sent are the correct length but look like unitialised RAM, matching the data in the raw binary files.  This is what I would expect to see if the script was run against the main Hacktribe branch, or if the read address is somehow getting mangled during transmission.

Can you confirm that you only have one Hacktribe connected via USB when you run the script?
Are you able to test using a Linux system?  Maybe mido interacts with the MIDI drivers differently on a Mac.

**Reply by AndrewSosna** [reply]
As it appeared, my SD card already contained a folder for Sampler version with e2sample.all so E2 never decompressed PCMs to memory and I was getting empty memory dump. I removed the SD card before dumping PCM samples and everything went fine, I got the proper WAV. Thank you very much.

Two following questions: 
1. Is there some document about the structure of this WAV, to make it easier for me to split it automatically, or loading it to Audacity or something similar and just splitting it by silence is the way to go?
2. Now I'm curious, is there a way do the reverse? To update this WAV by replacing samples and pushing them back to E2 flash, basically getting a regular E2 synth version with updated samples?

**Reply by bangcorrupt** [reply]
> split it automatically

I'm working on format definitions similar to those in [e2_formats.py](https://github.com/bangcorrupt/hacktribe-editor/blob/main/e2_formats.py), along with a script to read the full PCM section from flash.  These will allow parsing the binary to a dictionary, including a list of offsets for each sample.  I think there is at least few bytes of zeroes  between each sample, so splitting by silence may work, but if you can wait a while I should be able to script it.

> replacing samples and pushing them back to E2 flash

I don't know the compression algorithm used, but reversing it could be possible.  If I can't get the remaining features from the synth working in Hacktribe I might look at loading custom samples to the synth.  In the long run I want to move all the content to the SD card in a single project file, maybe using the flash for alternate firmware apps.  

Thank you again for your continued support :smiley: !

**Reply by alexreisinger** [reply]
Hello i would like to ask if there is more information on the structure of the wav or a script that splits the audio into the individual samples already? I have loaded it into Audacity and tried to label the samples by silence but that did not work well, there seem to be a minimum of a few audio points (samples) between those samples that i see.
Thanks for the dedication again!

**Comment by Ghillermo** [comment]
Hi, iirc the old "turn synth into sampler" header change hack allowed us to get the samples too. What happens if I run the python patch script but just use it to change header on src file (without applying the hack patch) would commenting a few lines of code (and calculating the new hash) do the trick or will it brick?

If we can do this then we'd just export samples to SD, move them to 501 an on and use them.

**Reply by bangcorrupt** [reply]
It will prompt to install the synth firmware.  Also, sampler firmware won't read the multi-sampled synth PCM correctly.

**Reply by AndrewSosna** [reply]
@bangcorrupt , Could you please tell more about multi-sampling support on E2/E2s?
As I understand the Sampler doesn't support multi-sampling but the Synth does?
Looks like not all but only some PCM samples are multi-sampled on Synth. So, theoretically, what would be the best way to get multi-sampling, I guess the only way is to replace PCM samples on Synth, right? And are there, like, "hardcoded multisample-dedicated PCM slots" on Synth, so that we would be only able to replace those, or theoretically all the "slots" (or does it even work this way, are there "slots" or just a memory area?) could contain multi-sampled samples?

**Reply by bangcorrupt** [reply]
They both support mullti-sampling but the sampler doesn't implement it fully.  The sampler flash PCM section is the same format as the synth, but every sample is in a group of one.  On playback the sampler reads and uses the empty key and velocity group data,  I was able to map samples by editing values in CPU RAM.  I didn't look into it much further, but I think when the sampler loads PCM from flash it assumes each sample is its own group which leads to bugs with instrument indexing and slice data saving.  I think some day we should at least be able to load a map file into RAM after booting, hopefully it can be integrated into the existing sample loading.

**Comment by AndrewSosna** [comment]
Yes, I only have one E2… I don’t have Linux, but I could install it on a virtual machine… or on WindowsAlso I could try with SysEx Librarian, it’s an app that sends SysEx from files and records the response. I was able to send “dump pattern” with it, but i would need a SysEx file. I can also try with Apple native API CoreAudio, I don’t have experience with it but it would be interesting to try. Still I’ll need to read your code and replicate it with Swift and form SysEx messages same way as you do (first time I’m dealing with this stuff)…Before I try the options above, are there any specific requirements when dumping PCM? I mean, maybe i should have removed the SE card before switching E2 on?Sent from a mobile deviceOn 2 Jan 2023, at 17:52, bangcorrupt ***@***.***> wrote:﻿
The firmware you sent me matches the one I tested and when I run it I get the samples dumped correctly.  The audio files you sent are the correct length but look like unitialised RAM, matching the data in the raw binary files.  This is what I would expect to see if the script was run against the main Hacktribe branch, or if the read address is somehow getting mangled during transmission.
Can you confirm that you only have one Hacktribe connected via USB when you run the script?
Are you able to test using a Linux system?  Maybe mido interacts with the MIDI drivers differently on a Mac.

**Reply by bangcorrupt** [reply]
> remove the SD card before switching E2 on

That could be it, didn't think of it.  If it loads an e2sample.all from the SD card it won't decompress the PCM from flash, so it will never copy it to CPU RAM.  When it shows the sampler boot screen, do you see the loading bar move across the bottom?  The first half moves in steps, this is reading the compressed PCM from flash.  The second half moves smoothly, this is decompressing chunks and sending them to the DSP RAM.  If it shows a percentage loading, then it is reading from the SD Card.

**Comment by AndrewSosna** [comment]
I’m not at home, will check when I go back, but as far as I remember indeed there was a progress indicator loading, with percentages Sent from a mobile deviceOn 2 Jan 2023, at 18:16, bangcorrupt ***@***.***> wrote:﻿

remove the SD card before switching E2 on

That could be it, didn't think of it.  If it loads an e2sample.all from the SD card it won't decompress the PCM from flash, so it will never copy it to CPU RAM.  When it shows the sampler boot screen, do you see the loading bar move across the bottom?  The first half moves in steps, this is reading the compressed PCM from flash.  The second half moves smoothly, this is decompressing chunks and sending them to the DSP RAM.  If it shows a percentage loafing, the it is reading from the SD Card.

**Comment by f-galland** [comment]
Andrew, is it possible you can share your wav file with the rest of us?

Thanks a lot, glad you got it working!

**Reply by AndrewSosna** [reply]
I'm sorry, I don't really want to get into copyright troubles, that's why I didn't ask @bangcorrupt to just send me his extracted file. 
I could try to help you with extracting your samples by writing a more detailed step-by-step guide, if that would help

**Reply by f-galland** [reply]
Thanks! It was more of a lazy request trying to avoid risk bricking my electribe xD

I didn't consider the copyright issue, though.

Happy new year!

**Reply by richardarnezjames** [reply]
DId you ever make a step by step guide for this? I'm having issues extracting the synth PCM file

**Comment by richardarnezjames** [comment]
I've tried for two days to get this to work and no matter what I get the soft-bricked screen that's says SOFTWARE UPDATE and nothing else. someone email me the the PCM file please. 
my E2 Synth's mother board is no. 13, I can't take anymore chances. 

**staballeyog@gmail.com**

**Comment by ptowns** [comment]
Not sure if this us relevant or of any use... 
https://www.chickensys.com/products2/korgcreator/index.html

This is a software for making multisample samples for other earlier korg devices, wonder if its the same?

**Comment by positionhigh** [comment]
i got the e2-synth-pcm.wav extracted with linux (attempts in windows failed).

Now, is there a script to get them back into an .all file or does that currently not exist, (yet) ? Maybe i am missing something obvious ?

**Reply by inkapmb** [reply]
I guess there is no script for that for the moment, you can chop the .wav with a DAW and use [Oe2sSLE](https://github.com/JonathanTaquet/Oe2sSLE) to create the .all

**Reply by vanana-old-acc** [reply]
On flash I saw there's a file with something along the lines of "KORGE2PCMFile" file header around 0x80000 iirc.. But I couldn't find anything on the internet about it so thats something to explore
Edit: i realized that these are factory samples not waveforms

**Reply by F-Thom** [reply]
The idea behind the access to ROM files is for me not about the samples themselves but about the usage of them: on the paper, it would be cool to keep the RAM free for custom samples but have additionally access to the ROM waveforms not as samples ready to chop and load but more as additional instruments loaded into RAM only if in use in a given project.
I guess the flash is too slow to read on the fly but there could be a RAM buffer reserved to load the RAM samples once they are in use.
The maximum buffer size shall be 16x the longest ROM sample (16 tracks) but you can in the practice also allow only 4 or 6 tracks to use ROM samples to save RAM.
Doing that you have the best of both worlds: room for custom samples + functional access to the ROM samples of the E2 (or E2S too probably) + all waveforms and filters...
But it's a totally different implementation at the end of the day.

**Reply by inkapmb** [reply]
@vananasun The waveforms are included in the those factory sample, when you extract them with sysex the waveforms are included in the .wav

@F-Tom 

> but more as additional instruments loaded into RAM only if in use in a given project

Yes sur, its exactly whta i want to do

> The maximum buffer size shall be 16x the longest ROM sample (16 tracks) but you can in the practice also allow only 4 or 6 tracks to use ROM samples to save RAM.

You thinks it will be too big ? Do you know how the synth of the e2 currently work ?

**Reply by vanana-old-acc** [reply]
So to clarify, are the waveforms stored in the CPU ROM, or on the Macronix flash chip? Because if it's stored in flash, that's good news

**Reply by inkapmb** [reply]
> So to clarify, are the waveforms stored in the CPU ROM, or on the Macronix flash chip? Because if it's stored in flash, that's good news

According to @bangcorrupt in this [comment](https://github.com/bangcorrupt/hacktribe/discussions/77#discussioncomment-1968979) : 

>  When the PCM data is loaded from the flash, it's decompressed and sent to DSP RAM one word at a time. I can hijack this to copy PCM to CPU RAM, then dump it over sysex as a single blob.

Therefore I guess the PCM data are natively decompressed and sent to the DSP RAM for the ROM (?), but this process have been modified in a special [branch](https://github.com/bangcorrupt/hacktribe/tree/synth-pcm-dump) to load PCM data to the CPU RAM to be able to dump them via sysex.

**Reply by F-Thom** [reply]
> Do you know how the synth of the e2 currently work ?
@inkapmb : unfortunately not but I can only guess: Korg will never integrate more RAM that it  would need. Now there is DSP RAM and CPU RAM.
The DSP RAM is filled with waveforms and this cover the 240 s of samples we use. If we reserve a bit of the DSP RAM to load ROM samples "on demand" while choosing a ROM sample as a sample.  
> You thinks it will be too big ?
I think the more we reserve, the less room remains for own samples to load or to sample/resample. Will any user need more than 6 ROM sample is a good question. If you decide to load 16 ROM sample, then you don't need Hacktribe at all, you can simply use the E2 stock firmware... But it may depend on the pattern you create.

A question remains though: if the ROM samples are compressed and loaded all together, it's difficult to access them, you need to decompress everything, get the sample you want, knowing the address and the length and dismiss the rest. Meanwhile, for this operation, you'll never have enough RAM, neither DSP or CUP... So you'll need to dismiss on the fly what you don't need and I think it's complicated to get.
Getting this working would be a dream but I don't consider this really feasible.

**Reply by inkapmb** [reply]
@F-Thom 

> If you decide to load 16 ROM sample, then you don't need Hacktribe at all, you can simply use the E2 stock firmware...

Sur but when switch to another pattern or switching synth/sample will get latency of loading ?

> A question remains though: if the ROM samples are compressed and loaded all together, it's difficult to access them, you need to decompress everything, get the sample you want, knowing the address and the length and dismiss the rest. Meanwhile, for this operation, you'll never have enough RAM, neither DSP or CUP... So you'll need to dismiss on the fly what you don't need and I think it's complicated to get.
Getting this working would be a dream but I don't consider this really feasible.

I fully agree with that from what I know. However, to avoid taking up too much memory space with certain synth wave form, it could be (potentially) possible to add another parameter to synth such as [_shift_] + (_osc_) to be able to morph these wave forms instead of having them in memory. Ex : There is a lot of _vpm-sine_, _vpm-saw_, etc... (from hack tribe) so it could be possible to one _vpm-sine_, one _vpm-saw_ etc, consequently it will be lot of free memory. In addition, according to the e2 [doc](https://cdn.korg.com/us/support/download/files/b77b1482084a8677d7f5ba24efc6cd5e.pdf) (p11), there is only 134 waveforms (from 246 to 380) rest being sample. I guess, _hack tribe synth_ (modified) + _e2 synth_ + _sample_ (with decent size) could fit, but I did not have dig enough in the firmware to confirm that.
